Fix string face

yuutayamada opened the pull request:
This PR is improved #44 with some tests.
(If test is not enough, I can wait)

pigmej commented:
Awesome :)

Does everything work as expected ? No more unknown / wrong behaviours ?

yuutayamada commented:
> Does everything work as expected ? No more unknown / wrong behaviours ?

The character delimiter works almost I expected.
But when I input '^@'(control character) as character, I got flycheck's error
"invalid character constant".

The control characters' highlight are working, but should I omit
those highlight? I implemented it because according to
http://nim-lang.org/docs/manual.html#lexical-analysis-character-literals
the character type is single byte...

Also I have little bit concern about typing single quote because
I changed syntax-table of single quote and I seem this affected
electric-pair mode. (should nim-mode provide similar function
for single quote?)

reactormonk commented:
Does anyone use electric-pair-mode that could comment on that part? I don't personally use it (so far).

yuutayamada commented:
Sorry, I didn't notice that Nim can use following form.

```nim
echo """""foo"""""""""""
```

Above commits apply this string delimiter's form.

yuutayamada commented:
> Does anyone use electric-pair-mode that could comment on that part? I don't personally use it (so far).

electric-pair-mode is kind of smart self-insert-command. For example, insert a single quote when
after number for number type specification, otherwise insert two single quotes for character delimiter.
The number of quote varies depends on context.

reactormonk commented on tests/test-syntax.el:
```diff
@@ -54,12 +119,32 @@
-  "should highlight comment correctly after char of double quote"
-  (test-concat-dir "tests/syntax/string.nim")
-  '(((119 . 165) . font-lock-comment-face))))
+  "should not highlight numbers by string-face"
```
I don't like the idea of `not` tests. Could you just expand it to a few tests for `number-face`?

yuutayamada replied:
Sure. I changed the test.

reactormonk commented:
I don't think it's necessary to correctly handle control characters in the sources. If you use these instead of `\n` or `\x23`, your code is bad and you should feel bad.

yuutayamada commented:
> I don't think it's necessary to correctly handle control characters in the sources. If you use these instead of \n or \x23, your code is bad and you should feel bad.

I agree with you. So should I omit control characters highlight?

reactormonk commented:
You can keep it in if it doesn't make the code more complicated. But if it conflicts with anything (electric apparently), toss it.

yuutayamada commented:
I updated the character delimiter's regex. The test is still working and I seem 
highlighting is working well. (I just checked syntax highlight on test file.)
